pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,195 @@
+	stopCh chan struct{}
+
+	// notificationMutex is used to block incoming events during an
+	// OnSyncDone event being sent back to the controller
+	ClusterID             string
+	ClientSet             kubernetes.Interface
+	podInformer           cache.SharedIndexInformer
+	networkPolicyInformer cache.SharedIndexInformer
+
+	eventChanMutex *sync.Mutex
+	eventChan      chan Event
+}
+
+const AddEvent = "Added"
+const DeleteEvent = "Deleted"
+const UpdateEvent = "Updated"
+
+const NetworkPolicy = "np"
+const Pod = "pod"
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    string
+	ObjType string
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
+	}()
+}
+
+func (rw *RemoteCluster) GetPods() []interface{} {
+	return rw.podInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rw.networkPolicyInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan Event) {
+	rw.eventChanMutex.Lock()
+	rw.eventChan = eventChan
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) enqueueEvent(event *Event) {
+
+	// lock used as memory barrier to make sure eventChannel ref
+	// is synchronized between threads
+	rw.eventChanMutex.Lock()
+	if rw.eventChan != nil {
+		rw.eventChan <- *event
```

Comment:
Yeah I knew, I thought it wasn't a big deal since most of the content is also pointers to other objects in the end. 

I remember thinking about this, and picking the whole object instead of reference, but it doesn't make any sense, since, here we don't care about mutability, etc... I'll turn it around, thanks ;)